### PR TITLE
#161291016 Enable pagination support

### PR DESCRIPTION
**Pull request by KvNGCzA**
#### What does this PR do?
have get `api/v1/articles?page=number&limit=number` route working with pagination support.

#### Description of Task to be completed?
- add pagination support for fetch articles route
- add validator for pageNumber
- add middleware to fetch all articles route
- add tests for fetch articles route

#### How should this be manually tested?
- `npm start` to start your server
- `npm run db-init` to pouplate your database
- use your browser or postman to hit the `localhost:{PORT}/api/v1/articles` route

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#161291016](https://www.pivotaltracker.com/story/show/161291016)

#### Screenshots (if appropriate)
Fetching all articles 
screen shot 2018-11-12 at 12 51 07 pm

Error if page query and limit query are not integers
screen shot 2018-11-12 at 12 56 18 pm

Error if page query or limit query are less than 1
screen shot 2018-11-12 at 12 57 32 pm

Error if available page number is exceeded
screen shot 2018-11-11 at 10 23 01 am

Error if queries are not provided
screen shot 2018-11-13 at 11 52 40 am

#### Questions:
N/A

### CODE REVIEW

File: server/controllers/ArticleController.js [hunk 1]
```diff
@@ -69,9 +74,20 @@ class ArticleController {
...
+              status: 'failure',
+              message: 'available number of page(s) exceeded'
+            }
+          });
+        }
+        return res.status(200).json({
           status: 'success',
-          articles: result.rows
+          availablePages: pageCount,
+          articles: rows
```
Comment (mbilesanmi) [inline server/controllers/ArticleController.js]: In addition, you should also be returning the total no of articles, current page and num of articles on current page

File: server/controllers/ArticleController.js [hunk 2]
```diff
@@ -50,14 +50,19 @@ class ArticleController {
...
+   * @returns {object} - returns all articles
    */
   static fetchAllArticles(req, res) {
+    const { page } = req.params;
+    const limit = 10;
+    const offset = (+page - 1) * limit;
     Article
       .findAndCountAll({
+        offset,
+        limit,
```
Comment (mbilesanmi) [inline server/controllers/ArticleController.js]: Have you handled search?

File: server/controllers/ArticleController.js [hunk 3]
```diff
@@ -50,14 +50,19 @@ class ArticleController {
...
-   * @param {object} res
-   * @returns {void}
+   * @description controller to fetch all articles
+   * @param {object} req - request object
+   * @param {object} res - response object
+   * @returns {object} - returns all articles
    */
   static fetchAllArticles(req, res) {
+    const { page } = req.params;
+    const limit = 10;
```
Comment (mbilesanmi) [inline server/controllers/ArticleController.js]: Why are we hardcoding this?

File: server/controllers/ArticleController.js [hunk 4]
```diff
@@ -50,14 +50,19 @@ class ArticleController {
...
-   * @param {object} req
-   * @param {object} res
-   * @returns {void}
+   * @description controller to fetch all articles
+   * @param {object} req - request object
+   * @param {object} res - response object
+   * @returns {object} - returns all articles
    */
   static fetchAllArticles(req, res) {
+    const { page } = req.params;
```
Comment (mbilesanmi) [inline server/controllers/ArticleController.js]: Like @sulenchy said.

File: server/controllers/ArticleController.js [hunk 5]
```diff
@@ -50,14 +50,19 @@ class ArticleController {
...
+    const limit = 10;
+    const offset = (+page - 1) * limit;
```
Comment (mbilesanmi) [inline server/controllers/ArticleController.js]: Shouldn't this be coming in from the user?

File: server/controllers/ArticleController.js (hunk 2 above)
Comment (KvNGCzA) [reply server/controllers/ArticleController.js]: @mbilesanmi no I have not

File: server/controllers/ArticleController.js (hunk 1 above)
Comment (KvNGCzA) [reply server/controllers/ArticleController.js]: @mbilesanmi done

File: server/middlewares/ArticleValidation.js [hunk 6]
```diff
@@ -47,6 +47,61 @@ class ArticleValidation {
...
     sendFormattedError(req, res, next);
   }
+
+  /**
+   * This method validates the page query
+   * @param {object} req - The request object
+   * @returns {void}
+   */
+  static validatePageQuery(req) {
+    req.checkQuery(
```
Comment (mbilesanmi) [inline server/middlewares/ArticleValidation.js]: Why is the page and limit required? It should not be.

File: server/controllers/ArticleController.js (hunk 2 above)
Comment (mbilesanmi) [reply server/controllers/ArticleController.js]: When would this be done?

Comment (KvNGCzA) [reply server/controllers/ArticleController.js]: @mbilesanmi its a separate task on the pt-board. So when this pagination task is approved I will start on that task

**Pull request merged**